pkg/chunked/storage_linux.go

```diff
@@ -232,28 +241,30 @@ func convertTarToZstdChunked(destDirectory string, blobSize int64, iss ImageSour
 	chunked, err := compressor.ZstdCompressor(f, newAnnotations, &level)
 	if err != nil {
 		f.Close()
-		return nil, "", nil, err
+		return nil, "", "", nil, err
 	}
 
-	digester := digest.Canonical.Digester()
-	hash := digester.Hash()
-
-	if _, err := io.Copy(io.MultiWriter(chunked, hash), diff); err != nil {
+	digesterUncompressed := digest.Canonical.Digester()
+	if _, err := io.Copy(io.MultiWriter(chunked, digesterUncompressed.Hash()), diff); err != nil {
 		f.Close()
-		return nil, "", nil, err
+		return nil, "", "", nil, err
 	}
 	if err := chunked.Close(); err != nil {
 		f.Close()
-		return nil, "", nil, err
+		return nil, "", "", nil, err
 	}
 	is := seekableFile{
 		file: f,
 	}
-	return &is, digester.Digest(), newAnnotations, nil
+
+	compressedDigest := digesterCompressed.Digest()
```

Comment:
(A) … I think the whole-layer digest match should be enforced here, at the latest.

---

I would actually argue that it would be preferable to enforce the digest match first, and _then_ only start parsing the tar file. This implementation exposes the tar parser and the zstd:chunked creation code to untrusted potentially-hostile data, unlike ordinary non-chunked layer pulls.